modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/CurrencySnapshotCreator.scala

```diff
@@ -76,111 +91,146 @@ object CurrencySnapshotCreator {
-        }
+          currentEpochProgress = trigger match {
+            case EventTrigger => lastArtifact.epochProgress
+            case TimeTrigger  => lastArtifact.epochProgress.next
+          }
+          lastActiveTips <- lastArtifact.activeTips
+          lastDeprecatedTips = lastArtifact.tips.deprecated
 
-        (deprecated, remainedActive, accepted) = getUpdatedTips(
-          lastActiveTips,
-          lastDeprecatedTips,
-          acceptanceResult,
-          currentOrdinal
-        )
+          maybeLastDataApplication = lastArtifact.data
 
-        (height, subHeight) <- getHeightAndSubHeight(lastArtifact, deprecated, remainedActive, accepted)
-
-        artifact = CurrencyIncrementalSnapshot(
-          currentOrdinal,
-          height,
-          subHeight,
-          lastArtifactHash,
-          accepted,
-          acceptedRewardTxs,
-          SnapshotTips(
-            deprecated = deprecated,
-            remainedActive = remainedActive
-          ),
-          stateProof,
-          currentEpochProgress,
-          maybeNewDataState
-        )
-      } yield
-        CurrencySnapshotCreationResult(
-          artifact,
-          CurrencySnapshotContext(lastContext.address, snapshotInfo),
-          awaitingBlocks.toSet,
-          rejectedBlocks.toSet
-        )
+          maybeLastDataState <- (maybeLastDataApplication, maybeDataApplicationService).mapN {
+            case (lastDataApplication, (_, service)) =>
+              service
+                .deserializeState(lastDataApplication)
+                .flatTap {
+                  case Left(err) => logger.warn(err)(s"Cannot deserialize custom state")
+                  case _         => Applicative[F].unit
+                }
+                .map(_.toOption)
+                .handleErrorWith(err =>
+                  logger.error(err)(s"Unhandled exception during deserialization data application, fallback to empty state") >>
+                    none[DataState].pure[F]
+                )
+          }.flatSequence
+
+          maybeNewDataState <- (maybeDataApplicationService, maybeLastDataState).mapN {
+            case ((nodeContext, service), lastState) =>
+              implicit val context: L0NodeContext[F] = nodeContext
+              NonEmptyList
+                .fromList(dataBlocks.distinctBy(_.value.roundId))
+                .map(_.flatMap(_.value.updates))
+                .map { updates =>
+                  service
+                    .validateData(lastState, updates)
+                    .flatTap { validated =>
+                      logger.warn(s"Data application is invalid, errors: ${validated.toString}").unlessA(validated.isValid)
+                    }
+                    .map(_.isValid)
+                    .handleErrorWith(err =>
+                      logger.error(err)(s"Unhandled exception during validating data application, assumed as invalid").as(false)
+                    )
+                    .ifM(
```

Comment:
Aren't you changing semantics here? I mean previously if the data was invalid you did run the combine method but with empty list of updates. Now you will not run it at all if I'm not mistaken. I'm not sure if the combine method is not incrementing something or performing some other update even though the list of updates is empty e.g. which we are preventing right now.